python/ray/util/spark/cluster_init.py

```diff
@@ -0,0 +1,681 @@
+                    last_alive_worker_count = cur_alive_worker_count
+                    last_progress_move_time = time.time()
+                    _logger.info(
+                        "Ray worker nodes are starting. Progress: "
+                        f"({cur_alive_worker_count} / {self.num_worker_nodes})"
+                    )
+                else:
+                    if time.time() - last_progress_move_time > 120:
+                        _logger.warning(
+                            "Timeout in waiting for all ray workers to start. Started / Total "
+                            f"requested: ({cur_alive_worker_count} / {self.num_worker_nodes}). "
+                            "Please check ray logs to see why some ray workers failed to start."
+                        )
+                        return
+        else:
+            _logger.warning("Already connected to this ray cluster.")
+
+    def disconnect(self):
+        if self.ray_context is not None:
+            try:
+                self.ray_context.disconnect()
+            except Exception as e:
+                # swallow exception.
+                _logger.warning(
+                    f"An error occurred while disconnecting from the ray cluster: "
+                    f"{repr(e)}"
+                )
+            self.ray_context = None
+        else:
+            _logger.warning("Already disconnected from this ray cluster.")
+
+    def shutdown(self):
+        """
+        Shutdown the ray cluster created by the `init_ray_cluster` API.
+        """
+        if not self.is_shutdown:
+            if self.ray_context is not None:
+                self.disconnect()
+            try:
+                self._cancel_background_spark_job()
+            except Exception as e:
+                # swallow exception.
+                _logger.warning(
+                    f"An error occurred while cancelling the ray cluster background spark job: "
+                    f"{repr(e)}"
+                )
+            try:
+                self.head_proc.kill()
+            except Exception as e:
+                # swallow exception.
+                _logger.warning(
+                    f"An Error occurred during shutdown of ray head node: {repr(e)}"
+                )
+            self.is_shutdown = True
+
+    def __enter__(self):
+        self.connect()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.shutdown()
+
+
+def _convert_ray_node_options(options):
+    return [f"--{k.replace('_', '-')}={str(v)}" for k, v in options.items()]
+
+
+_RAY_HEAD_STARTUP_TIMEOUT = 40
+_BACKGROUND_JOB_STARTUP_TIMEOUT = 30
+_RAY_CLUSTER_STARTUP_PROGRESS_CHECKING_INTERVAL = 3
+
+
+def _init_ray_cluster(
+    num_worker_nodes=None,
+    total_cpus=None,
+    total_gpus=None,
+    total_heap_memory_bytes=None,
+    total_object_store_memory_bytes=None,
+    heap_to_object_store_memory_ratio=None,
+    head_options=None,
```

Comment:
maybe this has been discussed. have we considered reuse `ray.init` options? https://docs.ray.io/en/latest/ray-core/package-ref.html#ray-init

Reply:
This might be more user-friendly,
but how to map `ray.init` options to `ray start` options ?
Maintaining the conversion in ray spark module is a burden.